### PR TITLE
quick and dirty way of checking pre and post stack effect

### PR DESCRIPTION
**Pull request by Z-Shang**
As mentioned in [this issue](https://github.com/vstinner/bytecode/issues/68),
I've modified compute stack size to take in consideration of the prerequisite stack size for some ops (e.g. `BUILD_CONST_KEY_MAP`)

For these definitely failing cases:
image

Before:
image

After:
image

### CODE REVIEW

**Comment by MatthieuDartiailh** [comment]
This looks reasonable enough. I will give a more detailed read later and try to remember why NOP used to be a problem.

**Comment by Z-Shang** [comment]
> This looks reasonable enough. I will give a more detailed read later and try to remember why NOP used to be a problem.

Thank you 👍

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
Some quick nitpick I caught through this a second time. Next I need to sit and check that we do catch all opcodes (for all supported Python version). I wish CPython would provide more tooling to support this.

File: bytecode/instr.py
Comment (Z-Shang) [reply bytecode/instr.py:286]: sure, I'll take a look at how the updates in Python 3.9 affect this

File: bytecode/instr.py
Comment (Z-Shang) [reply bytecode/instr.py:310]: by changing `pushes_back` to a free function, this will only be used here.

File: bytecode/instr.py
Comment (Z-Shang) [reply bytecode/instr.py]: I've added the Python 3.9 new ops, but by hard coding doesn't feel right.
We may have to find a better way in the future

File: bytecode/tests/test_concrete.py
Comment (Z-Shang) [reply bytecode/tests/test_concrete.py]: done

File: bytecode/instr.py
Comment (MatthieuDartiailh) [inline bytecode/instr.py:285]: Should have a leading underscore, this is meant to be private. And it does not need to be a method, we could extract the opcode name once, and pass it to an independent function, but this should be fine.

File: bytecode/instr.py
Comment (MatthieuDartiailh) [inline bytecode/instr.py:310]: We may want to handle the ROT_ opcodes in the same way since I assume the interpreter crashes if there are not enough objects on the stack.

File: bytecode/tests/test_bytecode.py
Comment (MatthieuDartiailh) [inline bytecode/tests/test_bytecode.py:277]: I would like to see some extra tests in particular for corner cases like DUP_TOP.

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
Can you also rebase on master ? I will do my best to give this the thorough read it deserves next week.

File: bytecode/instr.py
Comment (MatthieuDartiailh) [inline bytecode/instr.py:55]: I believe the following are missing from the list of push back (You could use a tuple rather than a list by the way):
SET_ADD
LIST_APPEND
MAP_ADD
IMPORT_
MAKE_FUNCTION
FORMAT_VALUE

File: bytecode/instr.py
Comment (MatthieuDartiailh) [reply bytecode/instr.py:316]: UNPACK_* opcodes need to be handled separately since they consume one value and pushed multiple ones

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
I noticed a bunch of possible issues. Please have a look and try to add additional tests to cover those.

File: bytecode/instr.py
Comment (Z-Shang) [reply bytecode/instr.py:55]: it is quite tricky for `IMPORT_` OPs
as `IMPORT_NAME` pops 2 and pushes 1
`IMPORT_FROM` pops 0 and pushes 1
`IMPORT_STAR` simply pops 1 and pushes nothing.

I think it then makes sense to consider only `IMPORT_NAME`

**Comment by Z-Shang** [comment]
I'm just wondering, it is possible to get the upstream (id est the Python team) to support this as a part of the standard library (such as `dis.stack_effect`)...

File: bytecode/instr.py
Comment (MatthieuDartiailh) [reply bytecode/instr.py]: I agree for IMPORT_

**Comment by MatthieuDartiailh** [comment]
Once we consolidate that here I will try to reach to the Python developers to suggest something along those lines indeed.

**Comment by MatthieuDartiailh** [comment]
Changes look nice. I will give it another thorough read during the week and convince myself it will always work (some exception handling have been worrying me but since the tests are passing I think everything is fine and just need to go through the logic once again myself). Hopefully I will merge once I am done. Could you add an entry to the change log in the meantime ?

**Comment by Z-Shang** [comment]
> Changes look nice. I will give it another thorough read during the week and convince myself it will always work (some exception handling have been worrying me but since the tests are passing I think everything is fine and just need to go through the logic once again myself). Hopefully I will merge once I am done. Could you add an entry to the change log in the meantime ?

for the changelog, do I also have to update the version number?

**Comment by MatthieuDartiailh** [comment]
You can just keep adding below 0.12 since it has not been released.

File: bytecode/instr.py
Comment (MatthieuDartiailh) [inline bytecode/instr.py:64]: Going through the list of opcodes again LOAD_ATTR should be listed

File: bytecode/instr.py
Comment (MatthieuDartiailh) [inline bytecode/instr.py:318]: FOR_ITER needs also a special case since it adds to the stack based on TOS for a jump not taken and pop TOS when jumping

File: bytecode/cfg.py
Comment (MatthieuDartiailh) [inline bytecode/cfg.py:144]: It could be nice while we stabilize this to have a way to bypass this in case we get the computation wrong (using an extra keyword to` _compute_stack_size`)

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
Going through the opcodes again I found a couple of cases I missed. Sorry for that. Also since I feel we may still get it wrong it may be interesting to provide a way to skip this check through an extra keyword argument.

Thanks for your work on this !

File: bytecode/cfg.py
Comment (Z-Shang) [reply bytecode/cfg.py:144]: I've made it a kw only argument :)

File: bytecode/instr.py
Comment (Z-Shang) [reply bytecode/instr.py:318]: actually I don't think this needs special consideration since the stack effect of  `FOR_ITER` is push 1 when no jump and pop 1 when jump, since the stack effect of non-pushing backs are all returned as "pre effect", this should be just fine.

Comment (MatthieuDartiailh) [reply bytecode/instr.py:318]: But in the first case you need something of the stack so for your analysis to hold you would need a pre effect of -1 and a post effect of 2, no ?

Comment (Z-Shang) [reply bytecode/instr.py:318]: ah yea, that makes sense then, thank you for the clarification

File: bytecode/cfg.py
Comment (MatthieuDartiailh) [inline bytecode/cfg.py:143]: I would do the following to not have to pass check_pre_and_post to the instruction.
```python
effects = instr.pre_and_post_stack_effect(jump=True) if check_pre_and_post else (0, instr.stack_effect(jump=True)
taken_size, maxsize = update_size(*effects, size, maxsize)
```
and something similar for the jump False case.

File: bytecode/instr.py
Comment (MatthieuDartiailh) [inline bytecode/instr.py:317]: I dislike the idea of this function not doing what its name suggests see my above suggestion.

File: bytecode/tests/test_bytecode.py
Comment (MatthieuDartiailh) [inline bytecode/tests/test_bytecode.py:284]: This test could be parametrize for all other similar opcodes.

File: bytecode/instr.py
Comment (MatthieuDartiailh) [inline bytecode/instr.py:337]: Having a test for this case would be nice

File: bytecode/instr.py
Comment (MatthieuDartiailh) [inline bytecode/instr.py:332]: Having a test for this case would be nice

File: bytecode/tests/test_bytecode.py
Comment (MatthieuDartiailh) [inline bytecode/tests/test_bytecode.py:305]: This test could be parametrized too to test all rot.

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
Couple minor things, also tests and documentation could be updated. Given the number of iterations I am putting you through let me know if you want me to help with anything.

File: bytecode/cfg.py
Comment (Z-Shang) [reply bytecode/cfg.py:143]: yea, that was also my first thought, but it seems that the coverage test would fail, that's why I changed it to this way

File: bytecode/instr.py
Comment (Z-Shang) [reply bytecode/instr.py:337]: I think I'd like some help on constructing a test case for this one, it would be great if you can help :)

**Comment by MatthieuDartiailh** [comment]
Sorry I have been awfully busy these past weeks. I will try to add the test you asked me to.

**Comment by Z-Shang** [comment]
No worries, I'm also busy as hell these days :(

**Comment by MatthieuDartiailh** [comment]
This looks really good. I will add a bit to the documentation in the coming week and then merge. Sorry again for the slow process I just moved from the USA back to France and I have been quite busy as a consequence.

**Comment by Z-Shang** [comment]
Pas de souci et prends ton temps :)

**Comment by MatthieuDartiailh** [comment]
I rebased to get tests running on 3.9. Once CIs run I will merge shortly. Thanks again for your contribution !

**Review by MatthieuDartiailh — APPROVED** [review]

**Pull request merged**
